language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+                     (TyForall ann out (Type ann) $
+                       -- the constructors' types with result-type replaced by out
+                       -- note to self: no need to normalize the individual constructor here;
+                       -- the outer,final,surrounding normalizeTypeM will reach and normalize each constructor
+                       foldr
+                       (TyFun ann . constructorCaseType (TyVar ann out))
+                       -- result type of constructor is: out
+                       (TyVar ann out)
+                       vdecls
+                     )
+
+       withVars :: Foldable t
+                => t (Name, Normalized (Type TyName uni ()))
+                -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+       withVars names m1 = foldr (uncurry withVar) m1 names
+
+-- | check that a binding is well-formed (correctly typed, kinded)
+checkWellformBind :: forall uni ann.
+                     (GShow uni, GEq uni, DefaultUni <: uni)
+                    => Recursivity
+                    -> Binding TyName Name uni ann
+                    -> TypeCheckM uni ann ()
+checkWellformBind recurs = \case
+  -- W-Term
+  TermBind _ _ (VarDecl ann _ ty) rhs -> do
+      checkKindM ann ty $ Type ()
+      -- OPTIMIZE: redundant, usually this function is called together with `withBind`,
+      -- which performs normalization here as well
+      vTy <- normalizeTypeM $ void ty
+      checkTypeM ann rhs vTy
+  -- W-Type
+  TypeBind _ (TyVarDecl ann _ k) rhs ->
+      checkKindM ann rhs (void k)
+  -- W-Data + W-Con
+  DatatypeBind _ (Datatype annD tvdecl@(TyVarDecl _ dataName _) tyargs des vdecls) -> do
+      -- Checks that tycons+tyargs are unique among eachother, and dataconstructors are unique among eachother
+      -- TODO: move these 2 checks to a different pass, e.g. in renaming phase
+      assertNoDuplicate allNewTyNames
+      assertNoDuplicate allNewNames
+      -- check the data-constructors' to be kinded by *
+      for_ vdecls $ \(VarDecl annV _ ty) -> do
+                  -- we normalize the type, since the user might have written some type-computation in the dataconstructor.
+                  -- we split the normalized type to a bunch of constructors arguments and a result-type.
+                  Normalized normTy <- normalizeTypeM ty
+                  let actualArgsTypes = funTyArgs normTy
+                      actualResType = funResultType normTy
+                  -- Check that the result-type is *-kinded, inside the whole env scope.
+                  -- The result-type can see all tyvardecls (tyargs + currently-defined type constructor).
+                  withTyVarDecls allNewTyDecls $
+                      checkKindM annV actualResType $ Type ()
+                  -- Check the dataconstructor arguments' types to also be *-kinded,
+                  -- but the environment defers for rec and nonrec bindings.
+                  withTyVarDecls
+                      (case recurs of
```

Comment:
I'd pull this out to a variable.

I think a good rule of thumb is: if you have a parenthesized expression that goes over more than one line, it should probably be a variable. If you can use `$` then it's usually okay.